### PR TITLE
Bump patch version of Rails to 5.2.8.1

### PR DESCRIPTION
**Pull request by mayorova**
**What this PR does / why we need it**:

This fixes a security issue described here: https://discuss.rubyonrails.org/t/cve-2022-32224-possible-rce-escalation-bug-with-serialized-columns-in-active-record/81017

**Which issue(s) this PR fixes** 

-

**Verification steps** 

-

**Special notes for your reviewer**:

-

### CODE REVIEW

**Review by jlledom — COMMENTED** [review]
I examined the [diff](https://github.com/rails/rails/compare/v5.2.8...v5.2.8.1) and it seems the only relevant change is [this](https://github.com/rails/rails/compare/v5.2.8...v5.2.8.1#diff-935494b14b54c907fc1d206871d2a4797c490b80f2c86d61a7264131dd492e42). And considering this is a patch, is supposed to not break anything. But this comment in the [CHANGELOG](https://github.com/rails/rails/compare/v5.2.8...v5.2.8.1#diff-353f4f101b1bb04938d761b6e29c88e0f147084e9e276854a77ae5d9a9a3b2a8) and the article you linked in the description suggest that there could be possible issues. Can we be sure we are not breaking anything?

**Comment by mayorova** [comment]
> Can we be sure we are not breaking anything?
Yeah, we need to verify it first :upside_down_face: 

I'll mark as draft for now.

**Comment by jlledom** [comment]
> We need to make sure usage of the affected class in porta is not affected:
> 
> ```
> 🐚 rg ActiveRecord::Coders::YAMLColum
> app/lib/event_store/event.rb
> 12:      YAML = ActiveRecord::Coders::YAMLColumn.new('EventStore', object_class = Hash)

If I understood this correctly, `Hash` is one of the classes considered safe so this wouldn't affect.

> And if needed allow the problematic classes like:
> 
> ```
> config.active_record.yaml_column_permitted_classes = [Symbol, Date, Time]
> ```

This something that we might need, yes.

> I don't see any usage of serialized classes in database. But that's also possible. One thing for example is #3226 where `PoliciesConfig` will have to be allowed.

Not sure what you mean by "serialized classes in database", but we have many columns which contain serialized objects formatted as yaml. e.g: `access_tokens.scopes` or all those models including `Fields` or `ExtraFields`.

**Comment by akostadinov** [comment]
Potentially we have such things but would you clarify? For example `scopes` appear to be fine locally.

```ruby
[3] pry(main)> AccessToken.pluck :scopes
   (1.1ms)  SELECT `access_tokens`.`scopes` FROM `access_tokens`
=> [["account_management"],
 ["account_management"],
 ["cms", "finance", "account_management", "stats", "policy_registry"],
 ["account_management"],
 ["account_management"],
 ["account_management"]]
```

**Comment by jlledom** [comment]
> Potentially we have such things but would you clarify? For example `scopes` appear to be fine locally.

Yes, I'm not saying it's going to fail. I just didn't understand what you meant when you said you didn't see any usage of serialized classes in database.

In the case of access tokens, those are stored in a column in yaml format, where all values are strings, so AIUI that shouldn't fail.

File: Gemfile [hunk 1]
```diff
@@ -12,7 +12,7 @@ gem 'aws-sdk-rails', '~> 3'
 gem 'aws-sdk-s3', '~> 1'
 
 gem 'dotenv-rails', '~> 2.7'
-gem 'rails', '~> 5.2.7'
+gem 'rails', '~> 5.2.8'
```
Comment (akostadinov) [inline Gemfile:15]: PR look good to me. Was changing this necessary? Not a problem. Just asking.

Comment (mayorova) [reply Gemfile:15]: It was not, but I thought it would be less confusing :wink:

File: test/unit/event_store/repository_test.rb [hunk 2]
```diff
@@ -1,3 +1,5 @@
+require 'test_helper'
```
Comment (mayorova) [inline test/unit/event_store/repository_test.rb:1]: This was not working locally without this.
I have no idea why it DID work in CircleCI...

Comment (jlledom) [reply test/unit/event_store/repository_test.rb:1]: Maybe because CircleCI runs this test along others? have you tried that locally?

File: test/unit/event_store/repository_test.rb [hunk 3]
```diff
@@ -28,7 +30,7 @@ def test_publish_event
   end
 
   def test_active_record_store
-    store = Repository.new(RailsEventStoreActiveRecord::EventRepository.new)
+    store = Repository.new(RailsEventStoreActiveRecord::EventRepository.new(adapter: EventStore::Event))
```
Comment (mayorova) [inline test/unit/event_store/repository_test.rb:33]: Took me a while to figure this one out... 

`EventStore::Event` implements serialization using GlobalId, where the objects are serialized by transforming into a global ID looking like `aj_globalid: gid://system/Account/28` (and later deserialized by fetching the object from the DB by ID).

Without this, the test was trying to serialize and deserialize the object Account (`provider` field in the event data), and this was failing with:

```
Tried to load unspecified class: Account
```

All the real events should be going through the repository with the correct adapter, so this is only a tests issue.

Comment (akostadinov) [reply test/unit/event_store/repository_test.rb:33]: nice investigation!

**Review by akostadinov — APPROVED** [review]

**Comment by mayorova** [comment]
The test `test_handles_errors` in `test/workers/send_user_invitation_worker_test.rb` was failing: https://app.circleci.com/pipelines/github/3scale/porta/23456/workflows/2270d5e1-5a46-4548-923f-b01729b1079b/jobs/264048

I spent some time trying to figure out why, but couldn't pinpoint the issue.

So, I resolved this by going back to the version `2.7.1` of the [`mail` gem](https://github.com/3scale/porta/pull/3251/commits/b34edf5c049052041c72e488b5a412b1c3c050ef#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR16), taking advantage that I had to rebase anyway to resolve conflicts with master.
Previously, it had upgraded to 2.8.0.
Probably, some change in that version is not compatible with the test (although I couldn't figure out what exactly). But also, it seems that there are some regressions, fixes for which are not yet released https://github.com/mikel/mail/blob/2-8-stable/CHANGELOG.rdoc#version-281-unreleased-

So, I think it's safer to keep the current version. Although, maybe somebody will be more successful in figuring out what exactly is wrong with this test, and fix it without locking the `mail` version.

File: test/unit/event_store/repository_test.rb (hunk 2 above)
Comment (mayorova) [reply test/unit/event_store/repository_test.rb:1]: Running the whole test suite? Hm, no...

Comment (jlledom) [reply test/unit/event_store/repository_test.rb:1]: No need to run the whole suit, only run it along any other test which includes the `require`:

```
[jlledom@fedora porta:patch-update-on-rails]$ bundle exec rake test:run TEST="test/unit/event_store/repository_test.rb"
Removing :environment prerequisite from db:create
TEST='
	test/unit/event_store/repository_test.rb
'
rake aborted!
NameError: uninitialized constant EventStore
/home/jlledom/redhat/porta/test/unit/event_store/repository_test.rb:1:in `<top (required)>'
/home/jlledom/.asdf/installs/ruby/2.6.8/bin/bundle:23:in `load'
/home/jlledom/.asdf/installs/ruby/2.6.8/bin/bundle:23:in `<main>'
Tasks: TOP => test:run => test
(See full trace by running task with --trace)


[jlledom@fedora porta:patch-update-on-rails]$ bundle exec rake test:run TEST="test/unit/event_store/*.rb"
Removing :environment prerequisite from db:create
TEST='
	test/unit/event_store/*.rb
'require
# Running tests with run options --seed 45192:

...............

Finished tests in 2.652558s, 5.6549 tests/s, 9.8019 assertions/s.


15 tests, 26 assertions, 0 failures, 0 errors, 0 skips
```

Comment (mayorova) [reply test/unit/event_store/repository_test.rb:1]: Thanks for trying this @jlledom !
Mystery solved then.

This require is still needed here IMO.

File: config/application.rb [hunk 4]
```diff
@@ -39,6 +39,16 @@ class Application < Rails::Application
...
     # Make `form_with` generate non-remote forms. Defaults true in Rails 5.1 to 6.0
+
+    # Applying the patch for CVE-2022-32224 broke YAML deserialization because some classes are disallowed in the serialized YAML
+    # NOTE: Symbol was later added to enabled classes by default, see https://github.com/rails/rails/pull/45584,
+    # it was added to Rails 6.0.6, 6.1.7, 7.0.4
+    config.active_record.yaml_column_permitted_classes = [Symbol, Time, Date, BigDecimal, OpenStruct,
+                                                          ActionController::Parameters,
+                                                          ActiveSupport::TimeWithZone,
+                                                          ActiveSupport::TimeZone,
+                                                          ActiveSupport::HashWithIndifferentAccess]
```
Comment (jlledom) [inline config/application.rb:50]: How did you come to this particular list of classes?

**Review by jlledom — APPROVED** [review]

File: config/application.rb (hunk 4 above)
Comment (mayorova) [reply config/application.rb:50]: Some details are in the PR description - basically just running the app locally, and then running the test suite, and seeing what fails :upside_down_face: 

There is a risk that some of the classes do not appear in the test suite, and it will fail for some edge case... I am not sure how to check that.

Comment (akostadinov) [reply config/application.rb:50]: Good point. I forgot about this list. We can grep an export of the database for YAML serialized objects.

**Review by thalesmiguel — APPROVED** [review]
